xclim/subset.py

```diff
@@ -125,11 +182,412 @@ def func_checker(*args, **kwargs):
 
     return func_checker
 
 
+def wrap_lons_and_split_at_greenwich(func):
+    @wraps(func)
+    def func_checker(*args, **kwargs):
+        """
+        A decorator to split and reproject polygon vectors in a GeoDataFram whose values cross the Greenwich Meridian.
+         Begins by examining whether the geometry bounds the supplied cross longitude = 0 and if so, proceeds to split
+         the polygons at the meridian into new polygons and erase a small buffer to prevent invalid geometries when
+         transforming the lons from WGS84 to WGS84 +lon_wrap=180 (longitudes from 0 to 360).
+
+         Returns a GeoDataFrame with the new features in a wrap_lon WGS84 projection if needed.
+        """
+        try:
+            poly = kwargs["poly"]
+            x_dim = kwargs["x_dim"]
+            wrap_lons = kwargs["wrap_lons"]
+        except KeyError:
+            return func(*args, **kwargs)
+
+        if wrap_lons:
+            if np.min(x_dim) < 0 or np.max(x_dim) > 360:
+                warnings.warn(
+                    "Dataset doesn't seem to be using lons between 0 and 360 degrees."
+                    " Tread with caution.",
+                    UserWarning,
+                    stacklevel=4,
+                )
+
+            if (poly.geometry.total_bounds[0] < 0) and (
+                poly.geometry.total_bounds[2] > 0
+            ):
+                warnings.warn(
+                    "Geometry crosses the Greenwich Meridian. Proceeding to split polygon at Greenwich."
+                    " This feature is experimental. Output might not be accurate.",
+                    UserWarning,
+                    stacklevel=4,
+                )
+
+                # Create a meridian line at Greenwich, split polygons at this line and erase a buffer line
+                union = Polygon(cascaded_union(poly.geometry))
+                meridian = LineString([Point(0, 90), Point(0, -90)])
+                buffered = meridian.buffer(0.000000001)
+                split_polygons = split(union, meridian)
+                buffered_split_polygons = [
+                    feat for feat in split_polygons.difference(buffered)
+                ]
+
+                # Load split features into a new GeoDataFrame with WGS84 CRS
+                split_gdf = gpd.GeoDataFrame(
+                    list(range(len(buffered_split_polygons))),
+                    geometry=buffered_split_polygons,
+                    crs={"init": "epsg:4326"},
+                )
+                # split_gdf.crs = CRS.from_epsg(4326)
+                split_gdf.columns = ["index", "geometry"]
+
+                poly = split_gdf
+
+            # Reproject features in WGS84 CSR to use 0 to 360 as longitudinal values
+            poly = poly.to_crs(
+                "+proj=longlat +ellps=WGS84 +lon_wrap=180 +datum=WGS84 +no_defs"
+            )
+            kwargs["poly"] = poly
+
+        return func(*args, **kwargs)
+
+    return func_checker
+
+
+@wrap_lons_and_split_at_greenwich
```

Comment:
Again, how does this compare with regionmask ?

Reply:
I haven't explored Regionmask too much, but it's essentially the same logic as Regionmask's `utils.wrap_angle`, except deployed as an automatic check with a wrapper.

Reply:
Does your implementation pass regionmask test suite ?

Reply:
I've looked into it now and the test approach for what I do here and the regionmask approach aren't exactly compatible. I do need to beef up the tests on this PR, but because the base type needed for the `create_mask` function is a GeoPandas dataframe, it's not all that similar to regionmask in how to define a new test region.